examples/introductory/algebra.wyv

```diff
@@ -0,0 +1,510 @@
+      () => simplifyProduct(t.u, t.v)
+    )
+    t:Pow => trySimplify1(t.u, uNew : Term => Pow(uNew, t.n),
+      () => simplifyPow(t.u, t.n)
+    )
+    t:Sin => trySimplify1(t.u, Sin,
+      () => NoStep
+    )
+    t:Cos => trySimplify1(t.u, Cos,
+      () => NoStep
+    )
+    t:Exp => trySimplify1(t.u, uNew : Term => Exp(t.b, uNew),
+      () => simplifyExp(t.b, t.u)
+    )
+    t:Log => trySimplify1(t.u, uNew : Term => Log(t.b, uNew),
+      () => simplifyLog(t.b, t.u)
+    )
+
+def fullySimplify(t : Term) : Term
+  match simplify(t):
+    a:Step => fullySimplify(a.term)
+    a:NoStep => t
+
+//------------------------------------------------------------------------------
+// Evaluation
+//------------------------------------------------------------------------------
+
+def eval(bindings : Map, t : Term) : Number = match t:
+  t:Constant => t.value
+  t:Variable => bindings.get(t.name).getOrElse(() => nan)
+  t:Sum => eval(bindings, t.u) + eval(bindings, t.v)
+  t:Product => eval(bindings, t.u) * eval(bindings, t.v)
+  t:Pow => realPow(eval(bindings, t.u), t.n)
+  t:Sin => sin(eval(bindings, t.u))
+  t:Cos => cos(eval(bindings, t.u))
+  t:Exp => realPow(t.b, eval(bindings, t.u))
+  t:Log => logBase(t.b, eval(bindings, t.u))
+
+//------------------------------------------------------------------------------
+// Examples
+//------------------------------------------------------------------------------
+
+val divider : String = "========================================"
+val divider2 : String = divider + divider
+
+val smallDivider : String = "-----"
+
+def printDivider() : Unit
+  stdout.print(divider2)
+  stdout.println()
+
+def printSmallDivider() : Unit
+  stdout.print(smallDivider)
+  stdout.println()
+
+def printInfo(t : Term, xVal : Number, yVal : Number) : Unit
+  val xString : String = numberToString(xVal)
+  val yString : String = numberToString(yVal)
+
+  var bindings : Map = map.make((x : String, y : String) => String.equals(x, y))
+  bindings.put("x", xVal)
+  bindings.put("y", yVal)
+
+  def printInfo1(title : String, s : Term) : Unit
+    val sNew = fullySimplify(s)
+
+    stdout.print("* " + title + ":\n    ")
+    stdout.print(toString(s))
+    stdout.print("\n  Fully simplified:\n    ")
+    stdout.print(toString(sNew))
+    stdout.print(
+      "\n  Evaluated @ (x = " + xString + ", y = " + yString + "):\n    "
+    )
+    stdout.print(numberToString(eval(bindings, sNew)))
+    stdout.print("\n")
+
+  printDivider()
+  printInfo1("Term", t)
+  printSmallDivider()
+  printInfo1("Derivative w.r.t. x", d("x", t))
+  printSmallDivider()
+  printInfo1("Derivative w.r.t. y", d("y", t))
+
+printInfo(
+  Sin(Pow(Variable("x"), 2.0)),
+  4.0, 5.0
+)
+printInfo(
+  Log(e, Exp(e, Pow(Variable("y"), 3.0))),
+  0.5, 2.0
+)
+printInfo(
+  Product(Variable("x"), Variable("y")),
+  3.0, 7.5
+)
+printInfo(
+  Sum(Pow(Sin(Variable("x")), 2.0), Pow(Cos(Variable("x")), 2.0)),
+  0.0, 0.0
+)
+
+printDivider()
```

Comment:
All very cool!  Do you mind adding a test case?  Even if you aren't checking the result, it is nice to know the examples will all continue to run without errors.